File: Sorting Techniques/Merge_Sort.cpp
#include<iostream>
using namespace std;

void merge(int arr[],int mid,int low,int high){
    int i=low;
    int j=mid+1;
    int B[100];
    int k=low;
    while(i<=mid && j<=high){
        if(arr[i]<arr[j]){
            B[k]=arr[i];
            i++;
            k++;
        }
        else{
            B[k]=arr[j];
            j++;
            k++;
        }
    }
    while(i<=mid){
        B[k]=arr[i];
        k++;
        i++;
    }
    while(j<=high){
        B[k]=arr[j];
        k++;
        j++;
    }
    for(int i=low;i<=high;i++){
        arr[i]=B[i];
    }

}

void Merge_Sort(int arr[],int low,int high){
    int mid;
    if(low<high){
        mid=(low+high)/2;
        Merge_Sort(arr,low,mid);
        Merge_Sort(arr,mid+1,high);
        merge(arr,mid,low,high);
    }
}

void printArray(int arr[],int n){
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}
int main(){
    int n;
    cout<<"Enter Size Of Array:";
    cin>>n;
    int arr[n];
    cout<<"Enter Array Elements:";
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    printArray(arr,n);
    Merge_Sort(arr,0,n-1);
    printArray(arr,n);
    return 0;
}

File: Recursion/MergeSort.java

import java.util.ArrayList;

public class MergeSort {

    public static void merge(int[] arr, int low, int mid, int high) {
        ArrayList<Integer> list = new ArrayList<>();
        int left = low;
        int right = mid + 1;

        while (left <= mid && right <= high) {
            if (arr[left] <= arr[right]) {
                list.add(arr[left]);
                left++;
            } else {
                list.add(arr[right]);
                right++;
            }
        }

        while (left <= mid) {
            list.add(arr[left]);
            left++;
        }

        while (right <= high) {
            list.add(arr[right]);
            right++;
        }

        for (int i = low; i <= high; i++) {
            arr[i] = list.get(i - low);
        }
    }

    public static void mergeSort(int[] arr, int low, int high) {
        if (low == high) {
            return;
        }
        int mid = (low + high) / 2;
        mergeSort(arr, low, mid);
        mergeSort(arr, mid + 1, high);
        merge(arr, low, mid, high);
    }

    public static void main(String[] args) {
        int[] arr = {3, 2, 4, 1, 3};
        int low = 0, high = arr.length - 1;

        for (int i = low; i <= high; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
        mergeSort(arr, low, high);

        for (int i = low; i <= high; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();

    }
}
